src/uct/tcp/tcp_listener.c

```diff
@@ -0,0 +1,196 @@
+/**
+* Copyright (C) Mellanox Technologies Ltd. 2019.  ALL RIGHTS RESERVED.
+*
+* See file LICENSE for terms.
+*/
+
+#include "tcp_listener.h"
+#include <ucs/sys/sock.h>
+#include <ucs/async/async.h>
+
+
+static void uct_tcp_listener_conn_req_handler(int fd, void *arg)
+{
+    uct_tcp_listener_t *listener = (uct_tcp_listener_t *)arg;
+    char ip_port_str[UCS_SOCKADDR_STRING_LEN];
+    struct sockaddr_storage client_addr;
+    ucs_async_context_t *async_ctx;
+    uct_tcp_sa_arg_t *sa_arg_ctx;
+    ucs_status_t status;
+    socklen_t addrlen;
+    int accept_fd;
+
+    ucs_assert(fd == listener->listen_fd);
+
+    async_ctx = listener->super.cm->iface.worker->async;
+
+    for (;;) {
+        addrlen = sizeof(struct sockaddr_storage);
+        status  = ucs_socket_accept(listener->listen_fd,
+                                    (struct sockaddr*)&client_addr,
+                                    &addrlen, &accept_fd);
+        if (status != UCS_OK) {
+            return;
+        }
+        ucs_assert(accept_fd != -1);
+
+        ucs_trace("server accepted a connection request from client %s",
+                  ucs_sockaddr_str((struct sockaddr*)&client_addr, ip_port_str,
+                                   UCS_SOCKADDR_STRING_LEN));
+
+        /* Set the accept_fd to non-blocking mode
+         * (so that send/recv won't be blocking) */
+        status = ucs_sys_fcntl_modfl(accept_fd, O_NONBLOCK, 0);
+        if (status != UCS_OK) {
+            goto err;
+        }
+
+        sa_arg_ctx = ucs_malloc(sizeof(uct_tcp_sa_arg_t), "accept sa_arg_ctx");
+        if (sa_arg_ctx == NULL) {
+            ucs_error("failed to allocate memory for a listen_ctx");
+            goto err;
+        }
+
+        sa_arg_ctx->fd = accept_fd;
+
+        UCS_ASYNC_BLOCK(async_ctx);
+        ucs_list_add_tail(&listener->sockcm->sa_arg_list, &sa_arg_ctx->list);
+        UCS_ASYNC_UNBLOCK(async_ctx);
+
+        status = ucs_async_set_event_handler(async_ctx->mode, accept_fd,
+                                             UCS_EVENT_SET_EVREAD |
+                                             UCS_EVENT_SET_EVERR,
+                                             uct_tcp_sa_data_handler,
+                                             sa_arg_ctx, async_ctx);
+        if (status != UCS_OK) {
+            goto err_free_ctx;
+        }
+    }
+
+err_free_ctx:
+    ucs_free(sa_arg_ctx);
+err:
+    close(accept_fd);
+}
+
+UCS_CLASS_INIT_FUNC(uct_tcp_listener_t, uct_cm_h cm,
+                    const struct sockaddr *saddr, socklen_t socklen,
+                    const uct_listener_params_t *params)
+{
+    ucs_async_context_t *async_ctx;
+    char ip_port_str[UCS_SOCKADDR_STRING_LEN];
+    ucs_status_t status;
+    int backlog;
+
+    UCS_CLASS_CALL_SUPER_INIT(uct_listener_t, cm);
+
+    self->sockcm          = ucs_derived_of(cm, uct_tcp_sockcm_t);
+    self->conn_request_cb = params->conn_request_cb;
```

Comment:
do we have field_mask bit for conn request?

Reply:
we have a bit for UCT_LISTENER_PARAM_FIELD_CONN_REQUEST_CB but it's already checked in uct_listener_create().